src/irc.py

```diff
@@ -0,0 +1,181 @@
+                f"An exception occurred when reloading {constants.IRC_BRIDGE_SNAP_NAME}. Reason: {e}"
+            )
+            logger.error(error_msg)
+            raise ReloadError(error_msg) from e
+
+    def start(self) -> None:
+        """Start the matrix-appservice-irc service.
+
+        Raises:
+            StartError: when encountering a SnapError
+        """
+        try:
+            cache = snap.SnapCache()
+            charmed_bind = cache[constants.IRC_BRIDGE_SNAP_NAME]
+            charmed_bind.start()
+        except snap.SnapError as e:
+            error_msg = (
+                f"An exception occurred when stopping {constants.IRC_BRIDGE_SNAP_NAME}. Reason: {e}"
+            )
+            logger.error(error_msg)
+            raise StartError(error_msg) from e
+
+    def stop(self) -> None:
+        """Stop the matrix-appservice-irc service.
+
+        Raises:
+            StopError: when encountering a SnapError
+        """
+        try:
+            cache = snap.SnapCache()
+            charmed_bind = cache[constants.IRC_BRIDGE_SNAP_NAME]
+            charmed_bind.stop()
+        except snap.SnapError as e:
+            error_msg = (
+                f"An exception occurred when stopping {constants.IRC_BRIDGE_SNAP_NAME}. Reason: {e}"
+            )
+            logger.error(error_msg)
+            raise StopError(error_msg) from e
+
+    def prepare(self) -> None:
+        """Prepare the machine."""
+        self._install_snap_package(
+            snap_name=constants.IRC_BRIDGE_SNAP_NAME,
+            snap_channel=constants.SNAP_PACKAGES[constants.IRC_BRIDGE_SNAP_NAME]["channel"],
+        )
+
+    def _install_snap_package(
+        self, snap_name: str, snap_channel: str, refresh: bool = False
+    ) -> None:
+        """Installs snap package.
+
+        Args:
+            snap_name: the snap package to install
+            snap_channel: the snap package channel
+            refresh: whether to refresh the snap if it's already present.
+
+        Raises:
+            InstallError: when encountering a SnapError or a SnapNotFoundError
+        """
+        try:
+            snap_cache = snap.SnapCache()
+            snap_package = snap_cache[snap_name]
+
+            if not snap_package.present or refresh:
+                snap_package.ensure(snap.SnapState.Latest, channel=snap_channel)
+
+        except (snap.SnapError, snap.SnapNotFoundError) as e:
+            error_msg = f"An exception occurred when installing {snap_name}. Reason: {e}"
+            logger.error(error_msg)
+            raise InstallError(error_msg) from e
+
+    def _write(self, path: pathlib.Path, source: str) -> None:
+        """Pushes a file to the unit.
+
+        Args:
+            path: The path of the file
+            source: The contents of the file to be pushed
+        """
+        with open(path, "w", encoding="utf-8") as write_file:
+            write_file.write(source)
+            logger.info("Pushed file %s", path)
```

Comment:
Would this be suitable to do `Path.write_text(source, encoding="utf-8")` here?

Reply:
I think so, nice one!